lib/spack/spack/util/file_cache.py

```diff
@@ -37,29 +38,29 @@ def __init__(self, root, timeout=120):
                 before assuming that there is a deadlock.
         """
         self.root = root.rstrip(os.path.sep)
-        if not os.path.exists(self.root):
+        if not Path(self.root).exists():
```

Comment:
What if we make some of these instance attributes Pathlib objects rather than strings and then create property handlers to control the casting to/from pathlib? That would cut down on some of the casting and allow for easier/more readable leveraging of the convenience of pathlib. This comment applies throughout this PR, as there are a number of places I think could stand for that kind of refactor.

Reply:
That makes sense. I took another pass over all of the files and tried to reduce the number of constructors and casts used. These changes look only change the internals to keep the class' api the same as before.